.github/workflows/code_analysis.yml

```diff
@@ -67,7 +67,28 @@ jobs:
       run: |
         files=$(cat changed_files.txt | grep '\.py$' || echo "")
         if [ -n "$files" ]; then
-          echo "$files" | xargs ruff format --diff
+          diff_command=""
+          apply_command=""
+          for file in $files; do
+            while IFS=- read -r start length; do
+              [ -z "$start" ] && continue
+              length=${length:-1}
+              end=$((start + length))
+              diff_command+="ruff format --diff --range $start-$end $file && "
+              apply_command+="ruff format --range $start-$end $file && "
+            done < <(git diff --unified=0 origin/$GITHUB_BASE_REF "$file" | grep '^@@' | sed -E 's/^@@ -[0-9]+(,[0-9]+)? \+([0-9]+)(,([0-9]+))? @@.*/\2-\4/')
+          done
+
+          if [ -n "$diff_command" ]; then
+            diff_command=${diff_command% && }
+            if ! eval "$diff_command"; then
+              apply_command=${apply_command% && }
+              echo "::error::Formatting failed. To apply the changes locally, run the following command: $apply_command"
```

Comment:
I would add a newline before `$apply_command` so it's more easily distinguishable from the error message